kotlinx-coroutines-core/common/src/channels/ConflatedBufferedChannel.kt

```diff
@@ -0,0 +1,177 @@
+    }
+
+    override suspend fun send(element: E) {
+        // Should never suspend, implement via `trySend(..)`.
+        val attempt = trySend(element)
+        if (attempt.isClosed) {
+            onUndeliveredElement?.callUndeliveredElement(element, coroutineContext)
+            throw sendException
+        }
+        assert { attempt.isSuccess }
+    }
+
+    override fun trySend(element: E): ChannelResult<Unit> = lock.withLock { // guard the operation by lock
+        // Is the buffer already full and the channel is not closed?
+        // In other words, should the plain `send(..)` operation on
+        // the buffer channel under the hood suspend?
+        if (!super.shouldSendSuspend()) {
+            // There is either a waiting receiver, space in the buffer,
+            // or this channel is closed for sending.
+            // Try to send the element into the buffered channel underhood.
+            return super.trySend(element).also {
+                assert { it.isClosed || it.isSuccess }
+            }
+        }
+        // The plain `send(..)` operation is bound to suspend.
+        // Either drop the latest or the older element, invoking
+        // `onUndeliveredElement` on it.
+        if (onBufferOverflow === DROP_LATEST) {
+            onUndeliveredElement?.invoke(element)
+        } else { // DROP_OLDEST
+            // Receive the oldest element. The call below
+            // does not invoke `onReceiveSynchronizationCompletion`.
+            val oldestElement = tryReceiveInternal().getOrThrow()
+            // Now, it is possible to send the element -- do it!
+            super.trySend(element).also {
+                assert { it.isSuccess }
+            }
+            // Finally, invoke `onUndeliveredElement`
+            // handler on the dropped oldest element.
+            onUndeliveredElement?.invoke(oldestElement)
+        }
+        return success(Unit)
+    }
+
+    override suspend fun sendBroadcast(element: E): Boolean {
+        trySend(element) // fails only when this channel is closed.
+            .onSuccess { return true }
+            .onClosed { return false }
+        error("unreachable")
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    override fun registerSelectForSend(select: SelectInstance<*>, element: Any?) {
+        // The plain `send(..)` operation never suspends. Thus, either this
+        // attempt to send the element succeeds or the channel is closed.
+        // In any case, complete this `select` in the registration phase.
+        trySend(element as E).let {
+            it.onSuccess {
+                select.selectInRegistrationPhase(Unit)
+                return
+            }.onClosed {
+                select.selectInRegistrationPhase(CHANNEL_CLOSED)
+                return
+            }
+        }
+        error("unreachable")
+    }
+
+    override fun shouldSendSuspend() = false // never suspends
+
+    override fun close(cause: Throwable?) = lock.withLock { // protected by lock
```

Comment:
I've added the following comment at the beginning of the class:
```
    // This implementation uses coarse-grained synchronization;
    // all channel operations are protected by this lock.
    private val lock = reentrantLock()
```

Let me know if a more detailed clarification is required.

Reply:
It would be really nice if the documentation could reflect the intent.
An example could be the following:
```
All channel operations are protected by this lock as the most straightforward way to support conflation on top of BufferedChannel -- all operations are performed under the lock in order to make the following conflation (with potential removal of the oldest element) atomic regarding other operations
```